### PR TITLE
Add Support for Guided Decoding to On Device Sampling

### PR DESCRIPTION
**Pull request by quic-sanising**
### 📌 Overview

This PR introduces **guided decoding** capabilities in On Device Sampling for `QEffForCausalLM` and `QEffCausalLMForTextImageToTextModel` models.

</br>
</br>

### 🚀 Motivation

As outlined in [this blog on structured decoding](https://blog.vllm.ai/2025/01/14/struct-decode-intro.html), structured decoding represents a fundamental shift in controlling LLM outputs. Instead of relying on post-processing, constraints are enforced during token generation via **logits manipulation**. This approach ensures:

*   **Format compliance** at generation time.
*   Reduced error rates for structured outputs.
*   Performance improvements through optimized backends like **XGrammar**, which can deliver up to **5× faster token generation under load**.

The constraints are provided through `token_bitmasks` which is a Boolean matrix of shape `(batch_size, vocab_size)`. Here, each element indicates whether a token should be kept (1) or masked (0). During sampling, this mask is applied to the logits before token selection, ensuring that only allowed tokens are considered. 

By performing this operation directly on the device, we eliminate host-device transfers, reduce latency, and improve throughput for structured decoding workloads.

</br>
</br>

### 🛠️ Implementation Details

The guided decoding logic is injected via `include_guided_decoding=True` during model loading. No changes to the model architecture are required.

```python
from QEfficient import QEFFAutoModelForCausalLM as AutoModelForCausalLM

# Load model with On Device Sampler enabled
qeff_model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-3.1-8B",
    continuous_batching=True,
    qaic_config={
        "include_sampler": True,
        "return_pdfs": False,
        "max_top_k_ids": 512,
        "include_guided_decoding": True,
    },
)

# Compile as usual
qeff_model.compile(
    prefill_seq_length=128,
    ctx_len=256,
    full_batch_size=16,
    num_devices=4,
    num_speculative_tokens=0,
    mxint8_kv_cache=True,
    mxfp6_matmul=True,
)
```

To disable guided decoding, simply set `include_guided_decoding=False`.

### CODE REVIEW

**Comment by quic-sanising** [comment]
Depends on #597

File: QEfficient/transformers/sampler/sampler.py [hunk 1]
```diff
@@ -122,6 +125,9 @@ def sampler_forward(
     top_ps: Optional[torch.Tensor] = None,
     min_ps: Optional[torch.Tensor] = None,
     random_numbers: Optional[torch.Tensor] = None,
+    vision_embeds: Optional[torch.Tensor] = None,
```
Comment (quic-mamta) [inline QEfficient/transformers/sampler/sampler.py]: please keep dtype of these 2 consistent as per lines 27-28. also update function docstring for these newly added args.

File: tests/transformers/sampler/test_sampler.py [hunk 2]
```diff
@@ -46,79 +75,162 @@
         20,  # generation_len
         4,  # full_batch_size
         1,  # spec_length
+        False,  # is_vlm
+    ),
+    # pytest.param(
```
Comment (quic-mamta) [inline tests/transformers/sampler/test_sampler.py]: Any reason for this test to be disabled?

File: tests/transformers/sampler/test_sampler.py [hunk 3]
```diff
@@ -132,55 +244,78 @@ def test_sampler_transform(
...
-    model_w_sampler = QEFFAutoModelForCausalLM.from_pretrained(
+    additional_configs = {}
+    additional_params = {}
+    if is_vlm:
+        config = AutoConfig.from_pretrained(model)
+        config.text_config.num_hidden_layers = 4
+        additional_configs["config"] = config
+        additional_configs["kv_offload"] = True
+        qeff_class = QEFFAutoModelForImageTextToText
+        assert isinstance(prompts, tuple)
```
Comment (quic-mamta) [inline tests/transformers/sampler/test_sampler.py]: Please add error message in this case for vlms.

File: tests/transformers/sampler/test_sampler.py [hunk 4]
```diff
@@ -233,48 +370,62 @@ def test_greedy_sampling(
...
+    if is_vlm:
+        additional_configs["kv_offload"] = True
+        qeff_class = QEFFAutoModelForImageTextToText
+        assert isinstance(prompts, tuple)
+        additional_params["images"] = prompts[0]
+        additional_params["processor"] = AutoProcessor.from_pretrained(model)
+        prompts = prompts[1]
+    else:
+        qeff_class = QEFFAutoModelForCausalLM
+        spec_length -= 1
```
Comment (quic-mamta) [inline tests/transformers/sampler/test_sampler.py]: num_hidden_layers should be specified here or does this test require all model layers?

File: tests/transformers/sampler/test_sampler.py [hunk 5]
```diff
@@ -385,3 +541,126 @@ def test_random_sampling(
...
+        },
+        **additional_configs,
+    )
+    model_w_sampler_wo_guided_decoding = qeff_class.from_pretrained(
+        model,
+        continuous_batching=True,
+        qaic_config={
+            "include_sampler": True,
+            "return_pdfs": False,
+            "max_top_k_ids": 1024,
```
Comment (quic-mamta) [inline tests/transformers/sampler/test_sampler.py]: can we define this test parameter in constants and then reuse?

File: QEfficient/transformers/models/pytorch_transforms.py [hunk 6]
```diff
@@ -711,10 +713,12 @@ class SamplerTransform:
         QEffGPTJForCausalLM,
         QEffGraniteForCausalLM,
         QEffGraniteMoeForCausalLM,
+        QEffInternDecoderWrapper,
         QEffLlamaForCausalLM,
         QEffMptForCausalLM,
         QEffPhi3ForCausalLM,
         QEffQwen2ForCausalLM,
+        QEffQwen_2_5_vl_DecoderWrapper,
```
Comment (quic-mamta) [inline QEfficient/transformers/models/pytorch_transforms.py]: Is this supported for Intern and Qwen models only?

Comment (quic-xiyushi) [reply QEfficient/transformers/models/pytorch_transforms.py]: Currently, only Intern and Qwen models have been tested. Other models should also be supported if we add their decoder wrapper here.

File: examples/performance/on_device_sampling.py [hunk 7]
```diff
@@ -21,13 +21,16 @@ def main(args, **kwargs):
     include_sampler = None
     return_pdfs = None
     max_top_k_ids = None
+    include_guided_decoding = None
```
[the remark this answers is not in the document]
Comment (quic-xiyushi) [reply examples/performance/on_device_sampling.py]: It’s set to None instead of a boolean because, in lines 46–54, include_guided_decoding will only be added to qaic_config if it is not None; otherwise, the default value defined in QEfficient/utils/sampler_utils.py will be used.

File: tests/transformers/sampler/test_sampler.py (hunk 5 above)
Comment (quic-xiyushi) [reply tests/transformers/sampler/test_sampler.py]: We want to use different `max_top_k_ids values`, and `include_guided_decoding` may or may not be present. Therefore, even if we define this test parameter in constants, it won’t be very reusable and is not generalized.

File: QEfficient/transformers/models/pytorch_transforms.py (hunk 6 above)
Comment (quic-xiyushi) [reply QEfficient/transformers/models/pytorch_transforms.py]: Added support for Llava, Garnite, Gemma, and Llama 4, and verified that they work correctly. Thanks!
